distmv/dirichlet.go

```diff
@@ -0,0 +1,142 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package distmv
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix/mat64"
+	"github.com/gonum/stat/distuv"
+)
+
+// Dirichlet implements the Dirichlet probability distribution.
+//
+// The Dirichlet distribution is a continuous probability distribution that
+// generates elements over the probability simplex, i.e. ||x||_1 = 1. The Dirichlet
+// distribution is the conjugate prior to the categorical distribution and the
+// multivariate version of the beta distribution. The probability of a point x is
+//  1/Beta(α) \prod_i x_i^(α_i - 1)
+// where Beta(α) is the multivariate Beta function (see the mathext package).
+//
+// For more information see https://en.wikipedia.org/wiki/Dirichlet_distribution
+type Dirichlet struct {
+	alpha []float64
+	dim   int
+	src   *rand.Rand
+
+	lbeta    float64
+	sumAlpha float64
+}
+
+// NewDirichlet creates a new dirichlet distribution with the given parameters alpha.
+// NewDirichlet will panic if len(alpha) == 0, or if any alpha is <= 0.
+func NewDirichlet(alpha []float64, src *rand.Rand) *Dirichlet {
+	dim := len(alpha)
+	if dim == 0 {
+		panic(badZeroDimension)
+	}
+	for _, v := range alpha {
+		if v <= 0 {
+			panic("dirichlet: non-positive alpha")
+		}
+	}
+	a := make([]float64, len(alpha))
+	copy(a, alpha)
+	d := &Dirichlet{
+		alpha: a,
+		dim:   dim,
+		src:   src,
+	}
+	d.lbeta, d.sumAlpha = d.genLBeta(a)
+	return d
+}
+
+// CovarianceMatrix returns the covariance matrix of the distribution. Upon
+// return, the value at element {i, j} of the covariance matrix is equal to
+// the covariance of the i^th and j^th variables.
+//  covariance(i, j) = E[(x_i - E[x_i])(x_j - E[x_j])]
+// If the input matrix is nil a new matrix is allocated, otherwise the result
+// is stored in-place into the input.
+func (d *Dirichlet) CovarianceMatrix(cov *mat64.SymDense) *mat64.SymDense {
+	if cov == nil {
+		cov = mat64.NewSymDense(d.Dim(), nil)
+	} else if cov.Symmetric() != d.dim {
+		panic("normal: input matrix size mismatch")
```

Comment:
I think the idea should work, though that, as written, doesn't work, right? We have to update the actual pointer itself.

````
if cov.Symmetric() == 0 {
     m := cov.GrowSquare(0, d.dim).(*mat64.SymDense) 
    *cov = *m
}

````